core/pyos.py

```diff
@@ -345,6 +346,27 @@ def InitInteractiveShell(self, display, my_pid):
         # For some reason setpgid() fails with Operation Not Permitted (EPERM) under pexpect?
         pass
 
+  def GetLastSignal(self):
+    # type: () -> int
+    """Return the last signal that fired"""
+    return self.last_sig_num
+
+  def GetHook(self, hook_name):
+    # type: (str) -> _TrapHandler
+    """Return the handler associated with hook_name"""
+    return self.hooks.get(hook_name, None)
+
+  def AddUserHook(self, hook_name, handler):
```

Comment:
I think technically hooks have nothing to do with signals and concurrent shared state -- they are purely "user space" / main loop


If we want to be really "pure" we could put them in a separate class, which is only used by `builtin_trap`

But that can always be done later, if at all

Reply:
Oh I like this! Removes the need for that awkward temporary code-dup I commented on in `leaky_core`